File: packages/gavio-py/gavio/interceptors/pii/scanners/swift_bic.py
"""SWIFT/BIC scanner — context-gated (FinTech pack).

Matches an 8- or 11-character BIC only when explicitly labelled ``SWIFT``/``BIC``,
so ordinary 8-letter uppercase words never trigger a false positive.
"""

from __future__ import annotations

import re

from ..context import ScanContext
from ..match import PiiMatch
from ..scanner import PiiScanner

# Keyword classes keep the label case-insensitive while requiring an UPPERCASE
# code (real BICs are uppercase). Group 1 is the code.
_SWIFT = re.compile(
    r"\b(?:[Ss][Ww][Ii][Ff][Tt]|[Bb][Ii][Cc])(?:\s+[Cc]ode)?\s*[:#]?\s*"
    r"([A-Z]{6}[A-Z0-9]{2}(?:[A-Z0-9]{3})?)\b"
)


class SwiftBicScanner(PiiScanner):
    entity_type = "SWIFT_BIC"

    def scan(self, text: str, ctx: ScanContext) -> list[PiiMatch]:
        out: list[PiiMatch] = []
        for m in _SWIFT.finditer(text):
            idx = ctx.next_index(self.entity_type)
            out.append(
                PiiMatch(
                    entity_type=self.entity_type,
                    start=m.start(1),
                    end=m.end(1),
                    value=m.group(1),
                    replacement=f"[SWIFT_BIC_{idx}]",
                )
            )
        return out

File: packages/gavio-js/src/interceptors/pii/scanners/swift-bic.ts
/** SWIFT/BIC scanner — context-gated (F-SEC-01, FinTech pack).
 *
 * Matches an 8- or 11-character BIC only when explicitly labelled `SWIFT`/`BIC`,
 * so ordinary 8-letter uppercase words never trigger a false positive.
 */

import type { ScanContext } from '../context.js'
import { makeMatch } from '../match.js'
import type { PiiMatch } from '../match.js'
import type { PiiScanner } from '../scanner.js'

// Keyword classes keep the label case-insensitive while requiring an UPPERCASE
// code (real BICs are uppercase). Group 1 is the code, at the end of the match.
const SWIFT =
  /\b(?:[Ss][Ww][Ii][Ff][Tt]|[Bb][Ii][Cc])(?:\s+[Cc]ode)?\s*[:#]?\s*([A-Z]{6}[A-Z0-9]{2}(?:[A-Z0-9]{3})?)\b/g

export function swiftBicScanner(): PiiScanner {
  return {
    entityType: 'SWIFT_BIC',
    tier: 1,
    scan(text: string, ctx: ScanContext): PiiMatch[] {
      const out: PiiMatch[] = []
      for (const m of text.matchAll(SWIFT)) {
        const code = m[1]!
        const end = m.index + m[0].length
        const start = end - code.length
        const idx = ctx.nextIndex('SWIFT_BIC')
        out.push(
          makeMatch({
            entityType: 'SWIFT_BIC',
            start,
            end,
            value: code,
            replacement: `[SWIFT_BIC_${idx}]`,
          }),
        )
      }
      return out
    },
  }
}
